collector/btrfs_linux.go

```diff
@@ -0,0 +1,194 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nobtrfs
+
+package collector
+
+import (
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs/btrfs"
+)
+
+// A btrfsCollector is a Collector which gathers metrics from Btrfs filesystems.
+type btrfsCollector struct {
+	fs     btrfs.FS
+	logger log.Logger
+}
+
+func init() {
+	registerCollector("btrfs", defaultEnabled, NewBtrfsCollector)
+}
+
+// NewBtrfsCollector returns a new Collector exposing Btrfs statistics.
+func NewBtrfsCollector(logger log.Logger) (Collector, error) {
+	fs, err := btrfs.NewFS(*sysPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open sysfs: %v", err)
+	}
+
+	return &btrfsCollector{
+		fs:     fs,
+		logger: logger,
+	}, nil
+}
+
+// Update retrieves and exports Btrfs statistics.
+// It implements Collector.
+func (c *btrfsCollector) Update(ch chan<- prometheus.Metric) error {
+	stats, err := c.fs.Stats()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve Btrfs stats: %v", err)
+	}
+
+	for _, s := range stats {
+		c.updateBtrfsStats(ch, s)
+	}
+
+	return nil
+}
+
+// btrfsMetric represents a single Btrfs metric that is converted into a Prometheus Metric.
+type btrfsMetric struct {
+	name            string
+	desc            string
+	value           float64
+	extraLabel      []string
+	extraLabelValue []string
+}
+
+// updateBtrfsStats collects statistics for one bcache ID.
+func (c *btrfsCollector) updateBtrfsStats(ch chan<- prometheus.Metric, s *btrfs.Stats) {
+	const subsystem = "btrfs"
+
+	// Basic information about the filesystem
+	devLabels := []string{"uuid"}
+
+	// Retrieve the metrics
+	metrics := c.getMetrics(s)
+
+	// Convert all gathered metrics to Prometheus Metrics and add to channel.
+	for _, m := range metrics {
+		labels := append(devLabels, m.extraLabel...)
+
+		desc := prometheus.NewDesc(
+			prometheus.BuildFQName(namespace, subsystem, m.name),
+			m.desc,
+			labels,
+			nil,
+		)
+
+		labelValues := []string{s.UUID}
+		if len(m.extraLabelValue) > 0 {
+			labelValues = append(labelValues, m.extraLabelValue...)
+		}
+
+		ch <- prometheus.MustNewConstMetric(
+			desc,
+			prometheus.GaugeValue,
+			m.value,
+			labelValues...,
+		)
+	}
+}
+
+// getMetrics returns metrics for the given Btrfs statistics
+func (c *btrfsCollector) getMetrics(s *btrfs.Stats) []btrfsMetric {
+	metrics := []btrfsMetric{
+		{
+			name:            "info",
+			desc:            "Filesystem information",
+			value:           1,
+			extraLabel:      []string{"label"},
+			extraLabelValue: []string{s.Label},
+		},
+		{
+			name:  "devices",
```

Comment:
This metric is extraneous, as it can be derived with a query like `count without (device) (node_btrfs_device_size_bytes)`. We like to avoid duplication of data.